### ISSUE TITLE
depending on a local target with a fully qualified label "@workspace_name//path/to:target" in repository rules (e.g. java_import_external) is considered external

### ISSUE THREAD

**Issue by natansil** [bazelbuild/bazel#6848]
### Description of the problem / feature request:

Depending on a **local** target with a fully qualified label "@local_workspace_name//path/to:target" in repository rules (e.g. `java_import_external`) is wrongfully considered **external**.

This issue reproduces both on bazel `HEAD` and on `0.21.0rc2 (2018-12-03)`

For example if a target `foo` depends on target `bar` , where both are in the workspace `baz_workspace` in the following manner
```
some_repository_rule(
   name = "foo",
   deps = ["@baz_workspace//:bar"],
)

some_rule(
   name = "bar"
)
```

then bar outputs **will reside in `external` area.**

This can cause serious issues for cyclic workspaces, where some labels are fully qualified and some are not.
for instance, depsets will not remove duplicates, which we already saw causes errors for rules that expect no duplicate files...

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I've reproduce this error with 2 different types of rules:

scala_maven_import_external:
https://github.com/natansil/remap_repro
```
bazel build //src/a_pkg:boo
...
DEBUG: /Users/natans/tmp/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file external/a/src/a_pkg/foo.jar>
```
java_import_external - https://github.com/natansil/remap_repro/tree/java_library (same target):
```
bazel build //src/a_pkg:boo
...
DEBUG: /Users/natans/tmp/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file external/a/src/a_pkg/libfoo.jar>
```

### What operating system are you running Bazel on?

macOS

### What's the output of `bazel info release`?

This issue reproduces on bazel `HEAD`, `0.21.0rc2 (2018-12-03)` and on bazel 0.19.0

@dkelmer @aehlig

**Comment by natansil** [comment]
FYI: @ittaiz @or-shachar

**Comment by allevato** [comment]
This also affects the Starlark debugger, because the Starlark interpreter will report different file paths (`external`-relative or workspace-relative) depending on how the `.bzl` file was loaded, and this affects breakpoint resolution: https://github.com/bazelbuild/vscode-bazel/issues/24

**Comment by natansil** [comment]
@dkelmer @aehlig do you have any updates on this issue? were you able to reproduce on your end?

**Comment by dkelmer** [comment]
If I understand the issue correctly, it seems like passing --experimental_remap_main_repo fixes this. Is the behavior observed with that flag the behavior you're expecting? 

If that does work, your next concern might be that this flag is "experimental" and that you shouldn't use experimental flags in production. I am going to both promote the flag from being experimental and make this the default behavior and in the meantime I would say that you should feel comfortable using this particular flag.

**Comment by natansil** [comment]
Thank you @dkelmer. you're right, I've forgot to add the experimental flag. my mistake.
There was still an issue in one of our repos that has this flag, that seemed to be related to the remap issue.
I'll try and get a small repro with the flag to try and understand where the problem resides.
One option is that at some point the dependencies are aggregated via list and not `depset` and so have this duplicate targets issue. I will double check.

Thanks again for your help!

**Comment by dkelmer** [comment]
If you could get a repro for the still failing case that would be great :)

**Comment by natansil** [comment]
@dkelmer I've managed to recreate the `external` mapping when the local fully qualified dependency resides in a repository rule. this time there is a .bazelrc file with the needed flag inside: `build --experimental_remap_main_repo=true`

Please pull the latest changes in both repo branches to see the repro of the failing case.

This case of dependency in repository rule is probably the root cause of the issue we're having in one of our "real" repositories at Wix.

**Comment by dkelmer** [comment]
Based on just reading the description of the problem, I expect you're running into this issue: https://github.com/bazelbuild/bazel/issues/6845. I'm confirming that it is indeed the same problem and that the fix at head also fixes your use case.

**Comment by dkelmer** [comment]
It doesn't fix your repro. I'll investigate further on Monday.

**Comment by dkelmer** [comment]
Ah, wait. I think it does fix the issue. I got thrown off by the artifacts that should actually be in external (like guava). 

The output with bazel 0.18 is (removing actual external artifacts):
```DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/foo.jar>
DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/non_qualified.jar>
DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/foo.jar>
DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file external/a/src/a_pkg/foo.jar>
DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/fully_qualified.jar>
```

And the output with bazel at HEAD is (removing actual external artifacts):
```DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/foo.jar>
DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/non_qualified.jar>
DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/foo.jar>
DEBUG: /Users/dannark/dev/bazel-workspaces/remap_repro/src/a_pkg/repro.bzl:7:17: dep: <generated file src/a_pkg/fully_qualified.jar>
```

Notably, the output with bazel at HEAD doesn't have any files from the external package and there is one less file because presumably it got deduped.

**Comment by natansil** [comment]
I'm happy to confirm that the issue we had in the "real" wix repo is gone when building with bazel at HEAD.
@dkelmer Thank you for the fix and your help!

**Comment by dkelmer** [comment]
Awesome, thanks for letting me know! I'm closing the issue.

**Comment by ittaiz** [comment]
Is this by any chance in any of the 0.21 RCs?

**Comment by ittaiz** [comment]
@natansil can you please check if 0.21 RCs work?

**Comment by natansil** [comment]
the fix is not part of 0.21
